### ISSUE TITLE
Not able to program (no response from microcontroller?)

### ISSUE THREAD

**Issue by bicyclesonthemoon** [msalau/rl78flash#8]
I am not able to use rl78flash for programming a device with a 104jg microcontroller.

2.b:~> rl78flash -ivvvvd -m3 /dev/ttyS0
		Open port: /dev/ttyS0
Using communication mode 3
Turn MCU's power on and press any key...
Send 1-byte data for setting mode
		send(1): 3A 
		recv(1): 3A 
Send "Set Baud Rate" command (baud=115200bps, voltage=3.3V)
		send(7): 01 03 9A 00 20 43 03 
		recv(7): 01 03 9A 00 20 43 03 
		recv(0): 
FAILED
Initialization failed
		Close port

It looks like the microcontroller is not responding.
That's exactly the same thing that happened here:
https://github.com/msalau/rl78flash/issues/6
In that case the problem was unconneced reset.

So I verified my circuit that it is electrically correct and it is.
This is how the signals look like (reset and tool0):
![sds00001](https://cloud.githubusercontent.com/assets/26207294/23591433/934eed3c-01f0-11e7-901f-4735990392d7.PNG)

And this is my circuit:
![flsch](https://cloud.githubusercontent.com/assets/26207294/23591539/06efa6ea-01f2-11e7-8def-7004a45139b3.PNG)

It's made after the rl78s-hw.png, but it has switches for selecting mode and power, an sp3232 chip and a 3v3 stabilizer. Powered from an usb cable.
I also have a second copy of this circuit made on a breadboard, without the switches. It's in mode 3. It doesn't work either.

It's also not the case of damaged microcontroller or device. I have access to several boards with a 104jg. None of them can be programmed with rl78flash but they all can be programmed with an E1 using the same cable.

So is the software doing something wrong? Like resetting  incorrectly? timing? whatever?
Or maybe I'm using it wrong somehow?

Thanks in advance,
B

**Comment by msalau** [comment]
Hi @bicyclesonthemoon 

The wave diagram looks good. The tool have send everything that is required by the protocol, but received no answer from the MCU (at least it is not shown on the diagram).

Do you really have to use the -d option? It is required only if utilize RESET pin as a GPIO. If this is not your case, please try without the option.

Since you have full control over the schematic, you can also try 2-wire UART mode (mode 2 and 4).

One more option to try is to increase [delay after de-assertion of RESET](https://github.com/msalau/rl78flash/blob/master/src/rl78.c#L70).

Regards,
Maksim.

**Comment by bicyclesonthemoon** [comment]
Thank you for the answer.
The -d option was not needed but it couldn't have broken anything.
2-wire mode is not something I would want to use because the device I'm trying to program has the txd and rxd lines connected to some additional circuits.

The increasing of delay after reset helped.
I increased it until it started working at 500. 

where did the original value of 10 come from?
I have seen that the E1, for example, waited 3ms after de-asserting reset.

**Comment by bicyclesonthemoon** [comment]
Unfortunately I still cannot program.
It gets stuck on this:
D:\b\rl78flash>rl78flash.exe -ewcvvv COM1 ex.srec
Send 1-byte data for setting mode
Send "Set Baud Rate" command (baud=115200bps, voltage=3.3V)
        OK
        Frequency: 32 MHz
        Mode: full-speed mode
Send "Reset" command
        OK
Send "Get Silicon Signature" command
        OK
        Device code: 100006
        Device name: R5F104JG
        Code flash size: 128kB
        Data flash size: 8kB
        Firmware version: 3.03
Erase
Send "Block Blank Check" command (range=000000..0003FF)
        OK
        Block is empty
Send "Block Blank Check" command (range=000400..0007FF)
        OK
        Block is empty
Send "Block Blank Check" command (range=000800..000BFF)
        OK
        Block is empty
Send "Block Blank Check" command (range=000C00..000FFF)
        OK
        Block is empty
Send "Block Blank Check" command (range=001000..0013FF)
FAILED
Block Blank Check failed (001000)
Read file "D11399_ASKO_WM7585_UI1.mot"
Write
Program block 000000
Send "Block Blank Check" command (range=000000..0003FF)
^C
D:\b\rl78flash>

It just gets stuck and doesn't do anything.
With -ewcvvvv I see this at the end:

srec_data (001FE0) : FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF
 FF FF FF FF FF FF FF FF FF FF FF FF
srec: S5031100EB

Record with no data (S5)
Write
Program block 000000
Send "Block Blank Check" command (range=000000..0003FF)
                send(12): 01 08 32 00 00 00 FF 03 00 00 C4 03
^C
D:\b\rl78flash>

The end of communication  looks like this:
![capture](https://cloud.githubusercontent.com/assets/26207294/23890565/b4e2e91e-0892-11e7-9a9d-96ff3eab0dde.PNG)
Here is my full capture:
3.672400062500000,0x3A
3.674516312500000,0x01
3.674611562500000,0x03
3.674706812500000,0x9A
3.674802062500000,0x00
3.674897000000000,0x20
3.674991812500000,0x43
3.675086625000000,0x03
3.678141937500000,0x02
3.678241125000000,0x03
3.678340250000000,0x06
3.678439375000000,0x20
3.678538500000000,0x00
3.678637687500000,0xD7
3.678736812500000,0x03
3.681427750000000,0x01
3.681523000000000,0x01
3.681618250000000,0x00
3.681713250000000,0xFF
3.681808062500000,0x03
3.681901687500000,0x02
3.681988687500000,0x01
3.682075562500000,0x06
3.682162500000000,0xF9
3.682249375000000,0x03
3.683785750000000,0x01
3.683880937500000,0x01
3.683976125000000,0xC0
3.684071375000000,0x3F
3.684166625000000,0x03
3.684256375000000,0x02
3.684343312500000,0x01
3.684430250000000,0x06
3.684517125000000,0xF9
3.684604062500000,0x03
3.684704062500000,0x02
3.684791000000000,0x16
3.684877875000000,0x10
3.684964812500000,0x00
3.685051687500000,0x06
3.685138625000000,0x52
3.685225500000000,0x35
3.685312500000000,0x46
3.685399375000000,0x31
3.685486312500000,0x30
3.685573187500000,0x34
3.685660187500000,0x4A
3.685747062500000,0x47
3.685834000000000,0x20
3.685920875000000,0x20
3.686007812500000,0xFF
3.686094750000000,0xFF
3.686181687500000,0x01
3.686268625000000,0xFF
3.686355625000000,0x2F
3.686442500000000,0x0F
3.686529500000000,0x03
3.686616312500000,0x00
3.686703312500000,0x03
3.686790187500000,0x5F
3.686877125000000,0x03
3.690523250000000,0x01
3.690618062500000,0x08
3.690713250000000,0x32
3.690808500000000,0x00
3.690903750000000,0x00
3.690999000000000,0x00
3.691094125000000,0xFF
3.691188937500000,0x03
3.691283750000000,0x00
3.691378500000000,0x00
3.691473312500000,0xC4
3.691568437500000,0x03
3.691867375000000,0x02
3.691954375000000,0x01
3.692041250000000,0x06
3.692128187500000,0xF9
3.692215062500000,0x03
3.694350625000000,0x01
3.694445875000000,0x08
3.694540812500000,0x32
3.694635625000000,0x00
3.694730437500000,0x04
3.694825250000000,0x00
3.694920062500000,0xFF
3.695015312500000,0x07
3.695110562500000,0x00
3.695205750000000,0x00
3.695301000000000,0xBC
3.695395937500000,0x03
3.695694812500000,0x02
3.695781812500000,0x01
3.695868687500000,0x06
3.695955625000000,0xF9
3.696042500000000,0x03
3.698435937500000,0x01
3.698531187500000,0x08
3.698626437500000,0x32
3.698721625000000,0x00
3.698816562500000,0x08
3.698911375000000,0x00
3.699006125000000,0xFF
3.699100937500000,0x0B
3.699195812500000,0x00
3.699291000000000,0x00
3.699386187500000,0xB4
3.699481437500000,0x03
3.699780312500000,0x02
3.699867187500000,0x01
3.699954187500000,0x06
3.700041125000000,0xF9
3.700128062500000,0x03
3.702116250000000,0x01
3.702211312500000,0x08
3.702306062500000,0x32
3.702400875000000,0x00
3.702495687500000,0x0C
3.702590500000000,0x00
3.702685625000000,0xFF
3.702780875000000,0x0F
3.702876125000000,0x00
3.702971375000000,0x00
3.703066437500000,0xAC
3.703161250000000,0x03
3.703460062500000,0x02
3.703547000000000,0x01
3.703633937500000,0x06
3.703720937500000,0xF9
3.703807812500000,0x03
3.706054187500000,0x01
3.706149500000000,0x08
3.706244750000000,0x32
3.706339937500000,0x00
3.706435187500000,0x10
3.706530000000000,0x00
3.706624812500000,0xFF
3.706719625000000,0x13
3.706814375000000,0x00
3.706909375000000,0x00
3.707004562500000,0xA4
3.707099812500000,0x03
3.707287375000000,0x02
3.707374312500000,0x01
3.707461187500000,0x1B
3.707548187500000,0xE4
3.707635062500000,0x03

**Comment by msalau** [comment]
Hi @bicyclesonthemoon 

> The increasing of delay after reset helped.
> I increased it until it started working at 500. (Edited to add: on Windows I had to increase this to 1000).

Thanks! I'll double check manual/datasheet for limitations and fix this.

> 0x13 is sent but not received. It repeats every time.

Does this happen on Linux too?
I have only 'small' MCUs at hand (up to 64k of flash), so I had no chance to verify the tool with larger memory space (your device has 128k of flash).

Also there is a bug in line https://github.com/msalau/rl78flash/blob/master/src/main.c#L230
It should be
` rc = rl78_erase(fd, code_size, data_size);`
The fix will make it fail earlier: right after failed erase.

Thanks for reporting. I'll investigate the issue with 0x13.

Maksim.

**Comment by bicyclesonthemoon** [comment]
> Does this happen on Linux too?
I will check this later. I use Linux at home and Windows at work.

> The fix will make it fail earlier: right after failed erase.
Indeed. Now it fails after the erase and does not get stuck.

**Comment by msalau** [comment]
Please, try the [fix-issue-8](https://github.com/msalau/rl78flash/tree/fix-issue-8) branch.
It has 3 fixes (compared to master):

* Increase delay after reset;
* Check return code of the erase function;
* Explicitly disable XON/OFF flow control on Windows.

The last one, I hope, should fix the issue with missing 0x13.

**Comment by bicyclesonthemoon** [comment]
I can confirm the fix.
I was able to successfully program the microcontroller on Windows.
I will check it on Linux later.

**Comment by msalau** [comment]
Hi @bicyclesonthemoon 
Many thanks for feedback! I've merged those changes to the master branch and created a new release: [v0.5.5](https://github.com/msalau/rl78flash/releases/tag/v0.5.5)

Best regards
Maksim

**Comment by bicyclesonthemoon** [comment]
After a first test this seems to work correctly.
I programmed the mcu with rl78flash and verified it with E1.
I programmed the mcu eith E1 and verified it with rl78flash.
It worked correctly.

(I had to program it with -ewc, without -r and to verify it with -dc because the software on the mcu overwrites the data flash on first run)

Later I will prepare a special test file which requires the whole flash to be written, then I will be sure that this works.
